canton/community/ledger/ledger-api-core/src/main/scala/com/digitalasset/canton/platform/apiserver/services/admin/ApiPackageManagementService.scala

```diff
@@ -106,23 +109,67 @@ private[apiserver] final class ApiPackageManagementService private (
     // Triggering computation in `executionContext` as caller thread (from netty)
     // should not be busy with heavy computation
-    val result = for {
-      darArchive <- Using(new ZipInputStream(darFile.newInput())) { stream =>
-        darReader.readArchive("package-upload", stream)
-      }
-      dar <- darArchive.handleError(Validation.handleLfArchiveError)
-      packages <- dar.all
-        .traverse(Decode.decodeArchive(_))
-        .handleError(Validation.handleLfArchiveError)
-      _ <- engine
-        .validatePackages(packages.toMap)
-        .handleError(Validation.handleLfEnginePackageError)
+    for {
+      (dar, decodedDar) <- Future.fromTry(
+        for {
+          darArchive <- Using(new ZipInputStream(darFile.newInput())) { stream =>
+            darReader.readArchive("package-upload", stream)
+          }
+          dar <- darArchive.handleError(Validation.handleLfArchiveError)
+          decodedDar <- dar
+            .traverse(Decode.decodeArchive(_))
+            .handleError(Validation.handleLfArchiveError)
+          _ <- engine
+            .validatePackages(decodedDar.all.toMap)
+            .handleError(Validation.handleLfEnginePackageError)
+        } yield (dar, decodedDar)
+      )
+      _ <- validateUpgrade(decodedDar)
     } yield dar
-    Future.fromTry(result)
+  }
+
+  private def validateUpgrade(upgradingPackage: Dar[(Ref.PackageId, Ast.Package)])(implicit
+      loggingContext: LoggingContextWithTrace
+  ): Future[Unit] = {
+    val upgradingPackageId = upgradingPackage.main._1
+    logger.info(s"Uploading DAR file for $upgradingPackageId in submission ID ${loggingContext.serializeFiltered("submissionId")}.")
+    upgradingPackage.main._2.metadata.flatMap(_.upgradedPackageId) match {
+      case Some(upgradedPackageId) =>
+        for {
+          upgradedArchiveMb <- packagesIndex.getLfArchive(upgradedPackageId)
+          upgradedPackageMb <- Future.fromTry {
+            upgradedArchiveMb.traverse(Decode.decodeArchive(_))
+              .handleError(Validation.handleLfArchiveError)
+          }
+          _ <- {
+            logger.info(s"Package $upgradingPackageId upgrades package id $upgradedPackageId")
+            val upgradeCheckResult = Typecheck.typecheckUpgrades(upgradingPackage.main, upgradedPackageMb)
+            upgradeCheckResult match {
+              case Success(()) => {
+                logger.info(s"Typechecking upgrades for $upgradingPackageId succeeded.")
+                Future(())
+              }
+              case Failure(err: UpgradeError) => {
+                logger.info(s"Typechecking upgrades for $upgradingPackageId failed with following message: ${err.message}")
+                Future.failed(Validation.handleUpgradeError(upgradingPackageId, upgradedPackageId, err).asGrpcError)
+              }
+              case Failure(err: Throwable) => {
+                logger.info(s"Typechecking upgrades failed with unknown error.")
+                Future.failed(err)
```

Comment:
Let's use a self-service error code for this as well: `PackageServiceError.InternalError.Unhandled` or define a custom one. In this way we also get the message logged at the correct logging level, which should probably error for this case

Reply:
fixed